### DISCUSSION TITLE
whats the correct way of extending cue with custom builtins ?

### DISCUSSION THREAD

**Discussion by cueckoo** [cue-lang/cue#1052] [Q&A]
_Originally opened by @ymolists in https://github.com/cuelang/cue/discussions/1052_

I am trying to see if i can build extensions for cue that dont need to live in the cue main repo. Something as a makefile replacement for example. All the make rules would be defined as cue structures. Its a toy example of course.

I am imagining that i would need to add a new set of builtins added. How would i go about ? anyone done this before ?

**Comment by cueckoo** [comment]
_Original reply by @myitcv in https://github.com/cuelang/cue/discussions/1052#discussioncomment-846565_

Please see discussion in https://github.com/cuelang/cue/issues/350

**Comment by mpvl** [comment]
Adding custom builtins is currently not supported: one needs to be really careful that computed results are hermetic. When it comes to generation, however, the `template` package is quite powerful.

That said, we do want to provide a way to write plugins at the "tooling" layer, possibly in WASM/JS or Go, for instance, that can be called from within CUE as an extensible mechanism.

If you feel you have functionality that belongs in the core builtins, that can be added of course.

**Reply by ningan123** [reply]
Hey, I wanna ask whether adding custom builtins is currently supported or not? If supported, how can I do it?

**Comment by mpvl** [comment]
@gedw99 If you write a Go program using the cue packages, it should be possible to cross compile to WASM/JS like any other Go program.

**Comment by mpvl** [comment]
@ningan123 We are inching towards having more support for this, and it is easier now to add custom builtins than it was before, but it is not supported yet in the public API.

Part of why it isn't there yet is lack of demand, part of this is that we have to watch out for a fragmented ecosystem, and part of that is that a more general Go/WASM/JS bridge in the tooling layer would possibly be a better solution, as it avoids the issue of adding non-hermetic functions (see #1522 on how easy it is to inadvertently add a non-hermetic function).

In general, if you have a function that you think is general enough, we can consider adding it as a builtins.

**Comment by jbcpollak** [comment]
I was just thinking of things that this would be useful for. Specifically, an annotation system that could be used to (for example) generate a UI editor based on a cue definition.
